docs/api.rst

```diff
@@ -0,0 +1,287 @@
+
+       If ``mutate`` is ``False``, then this does a deepcopy of the tree,
+       changes the copy, and returns the copy. This is expensive.
+
+   :arg boolean create_missing: If ``create_missing`` is ``False`` (the default),
+      then this will raise a ``KeyError`` for failed dict keys and
+      ``IndexError`` for failed list indexes.
+
+      If ``create_missing`` is ``True``, and this isn't
+      the last item in the path, then this will create the intermediary
+      dict/list.
+
+      If the next edge is a key, it'll create a dict. If the next edge is an
+      index, then it'll create a list filling in ``None`` for the required
+      indices.
+
+      Here are some examples.
+
+      This sets ``a`` to 5. This isn't affected by ``create_missing``.
+
+      >>> tree_set({}, 'a', value=5, create_missing=True)
+      {'a': 5}
+      >>> tree_set({}, 'a', value=5, create_missing=False)
+      {'a': 5}
+
+      This tries to traverse ``a``, but it doesn't exist and it's not the last
+      edge in the path. The next edge is ``b``, which is a key, so it first sets
+      ``a`` to an empty dict, then proceeds.
+
+      >>> tree_set({}, 'a.b', value=5, create_missing=True)
+      {'a': {'b': 5}}
+
+      This tries to traverse ``a``, but it doesn't exist and it's not the last
+      edge in the path. The next edge is ``[2]``, which is an index, so it first
+      sets ``a`` to a list of 3 ``None`` values, then proceeds.
+
+      >>> tree_set({}, 'a.[2]', value=5, create_missing=True)
+      {'a': [None, None, 5]}
+
+      This is similar, but with a negative index.
+
+      >>> tree_set({}, 'a.[-1]', value=5, create_missing=True)
+      {'a': [5]}
+
+      This creates missing indices in an existing list.
+
+      >>> tree_set({'a': []}, 'a.[2]', value=5, create_missing=True)
+      {'a': [None, None, 5]}
+
+
+   Examples:
+
+   These don't mutate the tree:
+
+   >>> tree = {'a': {'b': {'c': 1}}}
+   >>> tree_set(tree, 'a', value=5, mutate=False)
+   {'a': 5}
+   >>> tree_set(tree, 'a.b.c', value=[], mutate=False)
+   {'a': {'b': {'c': []}}}
+
+   These raise errors if an edge is missing:
+
+   >>> tree_set({}, 'a.b.c', value=5)
+   KeyError ...
+   >>> tree_set({}, 'a.[1].b', value=5)
+   IndexError ...
+
+   These create missing edges and indexes:
+
+   >>> tree_set({}, 'a.b.c', value=5, create_missing=True)
+   {'a': {'b': {'c': 5}}}
+   >>> tree_set({}, 'a.[1].b', value=5, create_missing=True)
+   {'a': [None, {'b': 5}]}
+
+
+.. py:func:: tree_flatten(tree)
+
+   Flattens a tree into a dict with keys of paths.
+
+   >>> tree_flatten({'a': 1})
+   {'a': 1}
+   >>> tree_flatten({'a': {'b': 1, 'c': 2}})
+   {'a.b': 1, 'a.c': 2}
+   >>> tree_flatten({'a': [{'b': 1}, {'c': 2}]})
+   {'a.[0].b': 1, 'a.[1].c': 2}
+
+   .. Note::
+
+      At this point, a flattened tree can't be used using ``tree_get`` and
+      ``tree_set``.
+
+
+.. py:func:: tree_setdefault(tree, default_tree)
+
+   FIXME
+
+
+.. py:func:: tree_validate(tree, schema)
```

Comment:
For your consideration, my favorite Python schema lib is https://pypi.python.org/pypi/schema/, by the docopt guy. I've done some work on it. I [use it in DXR](https://github.com/mozilla/dxr/blob/dd918bee56dd39a1cb8d220466da226ea569dc49/dxr/config.py#L208-L228). I've made some [design](https://github.com/keleshev/schema/issues/55) [proposals](https://github.com/keleshev/schema/pull/54) that I think would result in it being an elegant solution rather than being a bit slapdash in places.